### PR TITLE
DOC-152: Encryption and Decryption with multiple recipients using CMS

### PR DESCRIPTION
**Pull request by yatsenko-ihor**

### CODE REVIEW

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java [hunk 1]
```diff
@@ -94,6 +122,28 @@ public InputStream buildDecryptionInputStream(InputStream inputStream, Function<
         }
     }
 
+    @Override
+    @SneakyThrows
+    public InputStream buildDecryptionInputStream(InputStream inputStream, Key key, PublicKeyIDWithX509Cert x509Cert) {
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: This function should't exist - you don't know key before you open CMS envelope, it must be same lambda-function as is in buildDecryptionInputStream, but accepting list of key-names

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java [hunk 2]
```diff
@@ -51,6 +58,25 @@ public OutputStream buildEncryptionOutputStream(OutputStream dataContentStream,
         return streamEncrypt(dataContentStream, rec, encryptionConfig.getAlgorithm());
     }
 
+    @Override
+    @SneakyThrows
+    public OutputStream buildEncryptionOutputStream(OutputStream dataContentStream, List<PublicKeyIDWithX509Cert> publicKeyIDWithCertMap) {
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: This should accept PublicKeyWithId

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 3]
```diff
@@ -83,6 +83,26 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
         return result;
     }
 
+    @Override
+    @SneakyThrows
+    public List<PublicKeyIDWithX509Cert> getCertificates(KeyStoreAccess keyStoreAccess) {
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: It must be getPublicKeys function, don't duplicate it

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java (hunk 2 above)
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: Function buildEncryptionOutputStream on line 53 should simly accept list of PublicKey

File: datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CmsEncryptionServiceImplTest.java [hunk 4]
```diff
@@ -45,9 +44,43 @@
...
+        Security.addProvider(new BouncyCastleProvider());
         keyStoreAccess = getKeyStoreAccess();
     }
 
+    @Test
+    @SneakyThrows
+    public void testCmsStreamEnvelopeEncryptAndDecryptTestWithMultipleRecepients() {
+        KeyStoreAccess keyStoreAccess1 = getKeyStoreAccess();
+        KeyStoreAccess keyStoreAccess2 = getKeyStoreAccess();
+
```
Comment (electronicpeter) [inline datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CmsEncryptionServiceImplTest.java]: As both keystores have the same user and password this test does not proove too much. At least both of them should have different names and passwords. Further a third uses who is not in the list should try to decrypt too.

Comment (yatsenko-ihor) [reply datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CmsEncryptionServiceImplTest.java]: done

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java (hunk 2 above)
Comment (yatsenko-ihor) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: done

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 3 above)
Comment (yatsenko-ihor) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: done

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java (hunk 2 above)
Comment (yatsenko-ihor) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: done

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java (hunk 1 above)
Comment (yatsenko-ihor) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: done

**Comment by valb3r** [comment]
@yatsenko-ihor Declined in favor of:
https://github.com/adorsys/datasafe/pull/106
This branch has build failure because it is used from forked repository (when trying to deploy)

**Pull request closed without merging**